### PR TITLE
Port IISExpress tests to run on both IIS and IISExpress

### PR DESCRIPTION
**Pull request by jkotalik**
Part of organizing for #1005. We want to run IISExpress and IIS tests without having to call UseIISDeployer or have multiple InlineData calls. No actual code changes here.

### CODE REVIEW

**Comment by pakrym** [comment]
1. There are IISIntegration functional tests there so don't remove entire project
2. We would need 3 project structure eventually (common, IIS, IIS Express) so let's create it now.

**Comment by jkotalik** [comment]
Sure I can create the 3 projects with this PR.

**Comment by jkotalik** [comment]
There are a good set of tests that still haven't been ported to full IIS. I'll think it is better to get this PR in first, then fix the other issues (like log file creation, reusing published output, etc) in separate PRs.

**Comment by jkotalik** [comment]
Summary of changes:
- Rename IISIntegration.FunctionalTests to Common.FunctionalTests
- Add two test projects IISExpress.FunctionalTests and IIS.FunctionalTests
- Single skip condition [SkipIISTestCondition] which will vary between the IIS and IISExpress projects
- namespace IISIntegration.FunctionalTests to Microsoft.AspNetCore.Server.IIS.FunctionalTests.Utilities;
- Various flakiness fixes with the IIS deployer/application

File: test/IISExpress.FunctionalTests/InProcess/ShutdownTests.cs [hunk 1]
```diff
@@ -2,15 +2,14 @@
...
-using Microsoft.AspNetCore.Server.IntegrationTesting;
+using Microsoft.AspNetCore.Server.IIS.FunctionalTests.Utilities;
 using Microsoft.AspNetCore.Testing.xunit;
 using Xunit;
 using Xunit.Abstractions;
 
 namespace Microsoft.AspNetCore.Server.IISIntegration.FunctionalTests
 {
-    [SkipIfIISExpressSchemaMissingInProcess]
+    [SkipIISTestCondition]
```
Comment (pakrym) [inline test/IISExpress.FunctionalTests/InProcess/ShutdownTests.cs]: This needs a better name.

File: test/AspNetCoreModuleTests/AspNetCoreModuleTests.vcxproj [hunk 2]
```diff
@@ -1,178 +0,0 @@
-<?xml version="1.0" encoding="utf-8"?>
```
Comment (pakrym) [inline test/AspNetCoreModuleTests/AspNetCoreModuleTests.vcxproj]: Why was this deleted?

File: test/IISExpress.FunctionalTests/InProcess/ShutdownTests.cs (hunk 1 above)
Comment (jkotalik) [reply test/IISExpress.FunctionalTests/InProcess/ShutdownTests.cs]: SkipIfInstalltionInvalid?

File: test/AspNetCoreModuleTests/AspNetCoreModuleTests.vcxproj (hunk 2 above)
Comment (jkotalik) [reply test/AspNetCoreModuleTests/AspNetCoreModuleTests.vcxproj]: Unused project that should have been deleted a while ago. I could put it in a separate PR if you really want it to.

Comment (pakrym) [reply test/AspNetCoreModuleTests/AspNetCoreModuleTests.vcxproj]: Nope, if it's intentional it's fine.

File: test/IISExpress.FunctionalTests/InProcess/ShutdownTests.cs (hunk 1 above)
Comment (pakrym) [reply test/IISExpress.FunctionalTests/InProcess/ShutdownTests.cs]: RequiresIIS?

File: IISIntegration.sln [hunk 3]
```diff
@@ -380,6 +382,18 @@ Global
 		{1533E271-F61B-441B-8B74-59FB61DF0552}.Release|x64.Build.0 = Release|x64
 		{1533E271-F61B-441B-8B74-59FB61DF0552}.Release|x86.ActiveCfg = Release|Win32
 		{1533E271-F61B-441B-8B74-59FB61DF0552}.Release|x86.Build.0 = Release|Win32
+		{1F0C8D9B-F47B-41F3-9FC9-6954B6DC7712}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
```
Comment (pakrym) [inline IISIntegration.sln]: Are P2P dependencies between test projects and native set correctly?

File: test/Common.FunctionalTests/Utilities/IISApplication.cs [hunk 4]
```diff
@@ -177,17 +179,31 @@ private void ModifyAspNetCoreSectionInWebConfig(string key, string value)
...
+            if (File.Exists(_apphostConfigPath))
+            {
+                RetryHelper.RetryOperation(
+                    () => File.Delete(_apphostConfigPath),
+                    e => _logger.LogWarning($"Failed to delete file : {e.Message}"),
+                    retryCount: 3,
+                    retryDelayMilliseconds: 100);
+            }
 
             File.Move(_apphostConfigBackupPath, _apphostConfigPath);
```
Comment (pakrym) [inline test/Common.FunctionalTests/Utilities/IISApplication.cs]: Why some things are in retries and other aren't?

File: test/Common.FunctionalTests/Utilities/FunctionalTestsBase.cs [hunk 5]
```diff
@@ -25,14 +26,7 @@ public FunctionalTestsBase(ITestOutputHelper output = null) : base(output)
...
             // Currently hosting throws if the Servertype = IIS.
-            if (parameters.ServerType == ServerType.IIS)
-            {
-                _deployer = new IISDeployer(parameters, LoggerFactory);
-            }
-            else
-            {
-                _deployer = ApplicationDeployerFactory.Create(parameters, LoggerFactory);
-            }
+            _deployer = DeployerSelector.GetDeployer(parameters, LoggerFactory);
```
Comment (pakrym) [inline test/Common.FunctionalTests/Utilities/FunctionalTestsBase.cs]: Are we just silently ignoring parameters.Server?

Comment (jkotalik) [reply test/Common.FunctionalTests/Utilities/FunctionalTestsBase.cs]: Yeah, we have to if we wanted minimal code change between the two projects. I can remove setting .Server.

Comment (pakrym) [reply test/Common.FunctionalTests/Utilities/FunctionalTestsBase.cs]: Hm, what about exposing current server type as a static?

File: test/Common.FunctionalTests/Utilities/IISApplication.cs [hunk 6]
```diff
@@ -205,7 +205,12 @@ private void RestoreAppHostConfig()
                     retryDelayMilliseconds: 100);
             }
 
-            File.Move(_apphostConfigBackupPath, _apphostConfigPath);
+            RetryHelper.RetryOperation(
+                () => File.Move(_apphostConfigBackupPath, _apphostConfigPath),
+                e => _logger.LogError($"Failed to backup apphost.config: {e.Message}"),
+                retryCount: 3,
```
Comment (pakrym) [inline test/Common.FunctionalTests/Utilities/IISApplication.cs]: Add a helper, 3 and 100 are copied all over.

Comment (jkotalik) [reply test/Common.FunctionalTests/Utilities/IISApplication.cs]: I wanted to, but it felt weird making a helper of a helper 😅

File: test/Common.FunctionalTests/Utilities/FunctionalTestsBase.cs (hunk 5 above)
Comment (jkotalik) [reply test/Common.FunctionalTests/Utilities/FunctionalTestsBase.cs]: Sure, though I think we need to do something more. Is there a clean way to modify the test name to append the server type?

Comment (pakrym) [reply test/Common.FunctionalTests/Utilities/FunctionalTestsBase.cs]: Not that I know of.

File: IISIntegration.sln (hunk 3 above)
Comment (jkotalik) [reply IISIntegration.sln]: They depend on .IIS, which depends on the native components.

**Comment by jkotalik** [comment]
Previous failure in appveyor: https://github.com/aspnet/IISIntegration/issues/995

**Comment by jkotalik** [comment]
https://ci3.dot.net/job/aspnet_IISIntegration/job/release_2.2/job/windows-Configuration_Release_prtest/53/console was due to .deps.json being used by another process. Seems like a test parallelism issue?

**Comment by jkotalik** [comment]
@pakrym Added windows auth tests.

File: test/Common.FunctionalTests/AppHostConfig/IISExpress.config [hunk 7]
```diff
@@ -177,7 +177,7 @@
 
   <system.webServer>
 
-    <aspNetCore processPath="%LAUNCHER_PATH%" arguments="%LAUNCHER_ARGS%" hostingModel="[HostingModel]" stdoutLogEnabled="false" stdoutLogFile=".\logs\stdout" />
+    <aspNetCore processPath="%LAUNCHER_PATH%" arguments="%LAUNCHER_ARGS%" hostingModel="[HostingModel]" stdoutLogEnabled="false" stdoutLogFile=".\logs\stdout" forwardWindowsAuthToken="false" />
```
Comment (pakrym) [inline test/Common.FunctionalTests/AppHostConfig/IISExpress.config]: Why is forwardWindowsAuthToken disabled for IISExpress?

File: test/Common.FunctionalTests/OutOfProcess/HelloWorldTest.cs [hunk 8]
```diff
@@ -30,6 +31,9 @@ public static TestMatrix TestVariants
             {
                 ApplicationPath = Helpers.GetOutOfProcessTestSitesPath(),
             };
+            // The default in hosting sets windows auth to true.
+            // Set it to the IISExpress.config file
+            deploymentParameters.ServerConfigTemplateContent = File.ReadAllText("IISExpress.config");
```
Comment (pakrym) [inline test/Common.FunctionalTests/OutOfProcess/HelloWorldTest.cs]: Why not set it in the same place where deployer is selected?

File: test/Common.FunctionalTests/Utilities/IISFunctionalTestBase.cs [hunk 9]
```diff
@@ -18,7 +18,7 @@ public IISFunctionalTestBase(ITestOutputHelper output = null) : base(output)
 
         protected string GetServerConfig(Action<XElement> transform)
         {
-            var doc = XDocument.Load("AppHostConfig/Http.config");
+            var doc = XDocument.Load("IISExpress.config");
```
[the remark this answers is not in the document]
Comment (jkotalik) [reply test/Common.FunctionalTests/Utilities/IISFunctionalTestBase.cs]: There aren't any tests in Common, should we care?

[the remark this answers is not in the document]
Comment (pakrym) [reply test/Common.FunctionalTests/Utilities/IISFunctionalTestBase.cs]: But test using it won't work correctly on IIS because it'll pick up wrong base config file.

File: test/Common.FunctionalTests/OutOfProcess/HelloWorldTest.cs [hunk 10]
```diff
@@ -26,9 +27,12 @@ public static TestMatrix TestVariants
         [MemberData(nameof(TestVariants))]
         public async Task HelloWorld(TestVariant variant)
         {
+            // The default in hosting sets windows auth to true.
+            // Set it to the IISExpress.config file
             var deploymentParameters = new DeploymentParameters(variant)
             {
                 ApplicationPath = Helpers.GetOutOfProcessTestSitesPath(),
+                ServerConfigTemplateContent = File.ReadAllText("IISExpress.config")
```
[the remark this answers is not in the document]
Comment (jkotalik) [reply test/Common.FunctionalTests/OutOfProcess/HelloWorldTest.cs]: There is a Http.config that is used by default if the IISExpress.config isn't used. The only difference is the Http.config turns on windows auth by default. The reason I'm using the IISExpress config here is purely to have windows auth disabled. This also isn't used for IIS at all either. Cleanup needs to happen; I can make it so all tests use the IISExpress.config and only the tests that need windows auth use a modified config, or I can do it in a separate PR.

File: test/Common.FunctionalTests/Utilities/IISFunctionalTestBase.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (jkotalik) [reply test/Common.FunctionalTests/Utilities/IISFunctionalTestBase.cs]: You are right; the IIS tests currently ignore/ don't use the GetServerConfig method.

File: test/Common.FunctionalTests/OutOfProcess/HelloWorldTest.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (pakrym) [reply test/Common.FunctionalTests/OutOfProcess/HelloWorldTest.cs]: Do something like https://github.com/aspnet/IISIntegration/blob/e4b9fd0d340e8bb9c98b1c9afeb39c6ce632e74c/test/Common.FunctionalTests/Utilities/IISFunctionalTestBase.cs#L26 don't add another config just for one line change.

[the remark this answers is not in the document]
Comment (jkotalik) [reply test/Common.FunctionalTests/OutOfProcess/HelloWorldTest.cs]: I'm trying to move away from using the Http.config file in Hosting. I will make it so it only uses the IISExpress.config and modifies it to enable windows auth.
